addon/mixins/route.js

```diff
@@ -205,72 +194,98 @@ export default Ember.Mixin.create({
-            });
-          }
-          if (!this.get('_canLoadMore')) {
-            this.set(this.get('_modelPath') + '.reachedInfinity', true);
-            if(this.infinityModelLoaded) {
-              Ember.run.scheduleOnce('afterRender', this, 'infinityModelLoaded', {
-                totalPages: totalPages
-              });
-            }
-          }
-        },
-        () => {
-          this.set('_loadingMore', false);
-          throw new Ember.Error("Ember Infinity: Could not fetch Infinity Model. Please check your serverside configuration.");
-        }
-      );
-    } else {
-      if (!this.get('_canLoadMore')) {
-        this.set(this.get('_modelPath') + '.reachedInfinity', true);
-        if(this.infinityModelLoaded) {
-          Ember.run.scheduleOnce('afterRender', this, 'infinityModelLoaded', { totalPages: totalPages });
-        }
-      }
+    if (this.get('_loadingMore') || !this.get('_canLoadMore')) {
+      this._finishLoad();
+      return false;
     }
+
+    this.set('_loadingMore', true);
+
+    let promise = this._requestNextPage();
+
+    promise.then(
+      newObjects => {
+        this.updateInfinityModel(newObjects);
+        this.incrementProperty('currentPage');
+
+        this._notifyInfinityModelUpdated(newObjects);
+        this._finishLoad();
+      },
+      () => {
+        throw new Ember.Error("Ember Infinity: Could not fetch Infinity Model. Please check your serverside configuration.");
+      }
+    )
+    .finally(() => {
+      this.set('_loadingMore', false);
+    });
+
     return false;
   },
 
   /**
+   request the next page from the adapter
+
+   @method _requestNextPage
+   @returns {Ember.RSVP.Promise} A Promise that resolves the next page of objects
+   */
+  _requestNextPage() {
+    let nextPage    = this.get('currentPage') + 1;
+    let perPage     = this.get('_perPage');
+    let modelName   = this.get('_infinityModelName');
+
+    let requestPayloadBase = {};
+    requestPayloadBase[this.get('perPageParam')] = perPage;
+    requestPayloadBase[this.get('pageParam')] = nextPage;
+
+    let options = this._buildOptions();
+    let params = Ember.merge(requestPayloadBase, options);
+
+    return this.store[this._storeFindMethod](modelName, params);
+  },
+
+  /**
+   notify that the infinity model has been updated
+   @method _notifyInfinityModelUpdated
+   */
+  _notifyInfinityModelUpdated(newObjects) {
+    if (!this.infinityModelUpdated) {
+      return;
+    }
+
+    Ember.run.scheduleOnce('afterRender', this, 'infinityModelUpdated', {
+      lastPageLoaded: this.get('currentPage'),
+      totalPages: this.get('_totalPages'),
+      newObjects: newObjects
+    });
+  },
+
+  /**
+   finish the loading cycle by notifying that infinity has been reached
+
+   @method _finishLoad
+   */
+  _finishLoad() {
+    if (this.get('_canLoadMore')) {
+      return;
+    }
+
+    this.set(this.get('_modelPath') + '.reachedInfinity', true);
+    if(this.infinityModelLoaded) {
```

Comment:
`this.infinityModelLoaded` is always going to be undefined - it's not used anywhere else.  We should use `_canLoadMore`